### PR TITLE
Structured arrays support

### PR DESCRIPTION
**Pull request by aldanor**
This branch represents work-in-progress re: structured arrays support.

See `example20.cpp`, `example20.py` and `example20.ref` for the attached tests. Most of the implementation is contained in `numpy.h`.
- Added `py::memoryview` wrapper type with constructor from `buffer_info`.
- `py::array` constructor from buffer logic had to be rewritten to account for structured types.
  - Added `PYBIND11_NUMPY_DTYPE` macro for defining dtypes for structured arrays.
- `npy_format_descriptor::value` had to be changed to `npy_format_descriptor::dtype()` since the typenums are only sufficient for built-in types. Internally, it is converted to a descriptor anyway by numpy. However, this is backwards-compatible since the typenums for non-structured types are still accessible via `::value`.
- `format_descriptor::value` had to be changed to `format_descriptor::format()` so as to be able to use numpy's logic to general buffer format strings at runtime. However, the change is backwards compatible since the format strings known at compile time are still accessible via `::value`.

### CODE REVIEW

**Comment by aldanor** [comment]
Fixed segfault with nested descriptors, apparently you need to incref them before passing on to numpy API, the tests actually pass fine now.

File: include/pybind11/numpy.h [hunk 1]
```diff
@@ -98,31 +121,43 @@ class array : public buffer {
...
+        // Allocate zeroed memory if it hasn't been provided by the caller.
+        // Normally, we could leave this null for NumPy to allocate memory for us, but
+        // since we need a memoryview, the data pointer has to be non-null. NumPy uses
+        // malloc if NPY_NEEDS_INIT is not set (in which case it uses calloc); however,
+        // we don't have a descriptor yet (only a buffer format string), so we can't
+        // access the flags. The safest thing to do is thus to use calloc.
+        auto buf_info = info;
+        if (!buf_info.ptr)
+            // always allocate at least 1 element, same way as NumPy does it
+            buf_info.ptr = std::calloc(std::max(info.size, 1ul), info.itemsize);
```
Comment (wjakob) [inline include/pybind11/numpy.h]: Do you see any way to provide a way to allocate uninitialized memory, which was what the previous implementation did? I would say that in almost all cases, the user will allocate a buffer and then fill it evaluations with some kind of numerical routine, so the initialization is just wasted.

also, should this be (size_t) 1 instead of 1ul in the std::max call?

File: include/pybind11/common.h [hunk 2]
```diff
@@ -225,6 +225,31 @@ struct buffer_info {
     ~buffer_info() {
         if (view) { PyBuffer_Release(view); delete view; }
     }
+
+    Py_buffer& as_pybuffer() const {
```
Comment (wjakob) [inline include/pybind11/common.h]: I'd prefer to have this in the `memoryview` constructor, since it's breaking the pybind11 abstractions by exposing Python C API types. (and it's only ever used in `memoryview`)

Comment (aldanor) [inline include/pybind11/common.h]: Yes, this actually makes sense. I could move it there.

File: include/pybind11/numpy.h [hunk 3]
```diff
@@ -163,11 +216,121 @@ template <typename T> constexpr const int npy_format_descriptor<
...
 
+struct field_descriptor {
+    const char *name;
+    int offset;
+    object descr;
+};
+
+template <typename T>
+struct npy_format_descriptor<T, typename std::enable_if<is_pod_struct<T>::value>::type> {
+    static PYBIND11_DESCR name() { return _("user-defined"); }
```
Comment (wjakob) [inline include/pybind11/numpy.h]: better: `struct[...]` or something like that?

Comment (aldanor) [inline include/pybind11/numpy.h]: Hmm, well actually `PYBIND11_DTYPE` could pass the string version of the type (`#Type`) to format descriptor when registering a dtype, so we can print the actual type here, e.g. `"Foo"`, would that be better?

Comment (wjakob) [inline include/pybind11/numpy.h]: Sure, that makes sense. Just be careful that `name()` generally runs at compile time when using C++14 mode, so you can't do any runtime computation.

File: include/pybind11/numpy.h [hunk 4]
```diff
@@ -163,11 +216,121 @@ template <typename T> constexpr const int npy_format_descriptor<
...
+            names.append(str(field.name));
+            offsets.append(int_(field.offset));
+            formats.append(field.descr);
+        }
+        args["names"] = names;
+        args["offsets"] = offsets;
+        args["formats"] = formats;
+        // This is essentially the same as calling np.dtype() constructor in Python and passing
+        // it a dict of the form {'names': ..., 'formats': ..., 'offsets': ...}.
+        if (!api.PyArray_DescrConverter_(args.release().ptr(), &dtype_()) || !dtype_())
```
Comment (wjakob) [inline include/pybind11/numpy.h]: I feel like I'm missing something: where is dtype_() actually defined?

Comment (aldanor) [inline include/pybind11/numpy.h]: https://github.com/aldanor/pybind11/pull/1/files/6537e69d6072a3f0de507601e7571b13a56a23b2#diff-7174f0e1ad0a58b4622057815d7f8199R287

Comment (wjakob) [inline include/pybind11/numpy.h]: never mind!

File: include/pybind11/numpy.h [hunk 5]
```diff
@@ -163,11 +216,121 @@ template <typename T> constexpr const int npy_format_descriptor<
...
+                    return;
+                }
+            }
+        }
+        pybind11_fail("NumPy: failed to extract buffer format");
+    }
+
+private:
+    static inline PyObject*& dtype_() { static PyObject *ptr = nullptr; return ptr; }
+    static inline char* format_() { static char s[4096]; return s; }
```
Comment (wjakob) [inline include/pybind11/numpy.h]: The static char local variable seems rather hacky -- how about making it a class method? Also, dynamic allocation is probably the way to go here, wasting 4K for every dtype seems a bit overkill.

File: include/pybind11/numpy.h (hunk 1 above)
Comment (aldanor) [inline include/pybind11/numpy.h]: Re: `1ul`, yea, I guess I was just lazy :)

Re: malloc vs calloc, here's the only piece of code from numpy that refers to `NPY_NEEDS_INIT`:

``` cpp
    if (data == NULL) {
        /*
         * Allocate something even for zero-space arrays
         * e.g. shape=(0,) -- otherwise buffer exposure
         * (a.data) doesn't work as it should.
         */
        if (nbytes == 0) {
            nbytes = descr->elsize;
        }
        /*
         * It is bad to have uninitialized OBJECT pointers
         * which could also be sub-fields of a VOID array
         */
        if (zeroed || PyDataType_FLAGCHK(descr, NPY_NEEDS_INIT)) {
            data = npy_alloc_cache_zero(nbytes);
        }
        else {
            data = npy_alloc_cache(nbytes);
        }
        if (data == NULL) {
            PyErr_NoMemory();
            goto fail;
        }
        fa->flags |= NPY_ARRAY_OWNDATA;

    }
```

Realistically, if we never handle object dtypes, we can just use `malloc` instead of `calloc` I guess.

``` python
>>> a = np.array(['foo', 'bar'], dtype=object)
>>> memoryview(a).format
'O'
```

Or maybe we can check if there's any `'O'` in the format string, we switch to `calloc`.

File: include/pybind11/numpy.h (hunk 5 above)
Comment (aldanor) [inline include/pybind11/numpy.h]: What do you mean by making it a class method? This is a class method though; do you mean a static class variable? Yea, dynamic allocation we could do. Just occurred to me, what's if `PYBIND11_DTYPE` is called twice on the same type for some reason? Do we generally want to allow that or fail?

File: include/pybind11/numpy.h (hunk 1 above)
Comment (wjakob) [inline include/pybind11/numpy.h]: I think it's fine to ignore object types, which is somewhat esoteric in any case.

File: include/pybind11/numpy.h (hunk 5 above)
Comment (wjakob) [inline include/pybind11/numpy.h]: Sorry, I meant a class variable -- yes. I think it's entirely realistic that the same dtype might get registered multiple times. I guess this could be detected so that the second time becomes a no-op.

File: docs/advanced.rst [hunk 6]
```diff
@@ -1181,6 +1181,30 @@ template paramenter, and it ensures that non-conforming arguments are converted
 into an array satisfying the specified requirements instead of trying the next
 function overload.
 
+NumPy structured types
+======================
+
+In order for ``py::array_t`` to work with structured (record) types, we first need
+to register the memory layout of the type. This could be done via ``PYBIND11_DTYPE``
```
Comment (wjakob) [inline docs/advanced.rst]: minor: could->can

File: include/pybind11/numpy.h (hunk 5 above)
Comment (aldanor) [inline include/pybind11/numpy.h]: Not really:

``` cpp
struct A { int a; int b; };
PYBIND11_DTYPE(A, a);
PYBIND11_DTYPE(A, b);
```

This is completely vaild, don' have to refer to all fields.

File: example/example18.cpp [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+    auto to_str = [](py::object obj) {
+        return (std::string) (py::str) ((py::object) obj.attr("__str__"))();
+    };
+    std::cout << to_str(py::dtype_of<SimpleStruct>()) << std::endl;
+    std::cout << to_str(py::dtype_of<PackedStruct>()) << std::endl;
+    std::cout << to_str(py::dtype_of<NestedStruct>()) << std::endl;
+}
+
+void init_ex18(py::module &m) {
+    PYBIND11_DTYPE(SimpleStruct, x, y, z);
```
Comment (wjakob) [inline example/example18.cpp]: Since PYBIND11_DTYPE is entirely NumPy specific, I would actually suggest to rename it to PYBIND11_NUMPY_DTYPE

File: include/pybind11/numpy.h (hunk 5 above)
Comment (wjakob) [inline include/pybind11/numpy.h]: But I guess this would break things elsewhere (which might refer to field 'a')?

Comment (aldanor) [inline include/pybind11/numpy.h]: Well, it depends... In C++ you will still work with struct A, but in Python you would have structs that will either see the `a` part or the `b` part (a "view" of sorts). I can't see a particular use case for this though, and since we essentially glue the dtype (which is a runtime thing) to the C++ type (which is not a runtime thing), maybe it makes sense to ensure it's called at most once for each C++ type?..

Ha, btw I believe this may have pointed me at a potential bug, if you do a `PYBIND11_DTYPE(A, a)`, I think it will set itemsize incorrectly (because I forgot to pass `sizeof(A)`) to dtype initializer. I'll fix it and add some tests.

**Comment by aldanor** [comment]
@wjakob Hey, I think I've fixed all issues with padding and stuff, rebased onto the most recent master (this took me a good while since there were name clashes, had to rename example18.\* to example20.*, do some git filter-branch dark magic etc...), now just following up on the rest of your comments here and it will be then done :)

A few things: 
- Is there a great value in having `PYBIND11_DESCR` being constexpr (and hence `descr` not being constructible from runtime strings) other than it being fancy? I see it's only used to generate signatures in docstrings, that sounds quite runtime to me :)

It would be sure nice to see signatures like

``` python
def f(arr : numpy.ndarray[dtype=Foo]) -> numpy.ndarray[dtype=Bar]
```

instead of it just being `"struct"` for all user-defined types. Are there any workarounds around this?
- Do you think it's worth introducing a `py::dtype` object? I don't think your previous suggestion (i.e. have methods like `.append(offset, name, format)`) would work since not all dtypes are structured, so you would need a whole hierarchy of c++ classes. But maybe moving padding stripping logic and a few other things into `py::dtype` would make sense (e.g., construct dtype from buffer format string, extract it from an array object, etc). Also there's now functions like `py::dtype_of<T>()` (maybe makes sense to make this a static function in `py::dtype`?) and `py::array::dtype()` that should naturally return `py::dtype` objects.
- Re: static class variables, your suggestion wouldn't work, since they're not constexpr and this library is header-only, so there's no way to statically initialize them. Which implies they should be kept in those ugly static functions...

P.S. by the way there's a few exception leaks in examples 5 and 18 in master, I've posted the issue to the main repo

File: example/example20.cpp [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+} __attribute__((packed));
+
+std::ostream& operator<<(std::ostream& os, const PackedStruct& v) {
+    return os << "p:" << v.x << "," << v.y << "," << v.z;
+}
+
+struct NestedStruct {
+    SimpleStruct a;
+    PackedStruct b;
+} __attribute__((packed));
```
Comment (wjakob) [inline example/example20.cpp]: Don't think MSVC will like this attribute. Do you want to turn the PR into an official one at the pybind11 repository? That way, the CI build bots will try to compile your code on various platforms, which may uncover a few more things.

File: example/example20.cpp [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+py::array mkarray_via_buffer(size_t n) {
+    return py::array(py::buffer_info(nullptr, sizeof(T),
+                                     py::format_descriptor<T>::format(),
+                                     1, { n }, { sizeof(T) }));
+}
+
+template <typename S>
+py::array_t<S, 0> create_recarray(size_t n) {
+    auto arr = mkarray_via_buffer<S>(n);
+    auto ptr = static_cast<S*>(arr.request().ptr);
```
Comment (wjakob) [inline example/example20.cpp]: Better to save the return value of request() in a local variable. Otherwise the buffer is released before you fill it below. No functional difference in most cases, but I'd say it's better practice to be careful.

File: example/example20.cpp (hunk 8 above)
Comment (aldanor) [inline example/example20.cpp]: Oh, good catch, I'll fix that tonight and open a PR to pybind11 :)

// please see my other few comments above

File: include/pybind11/common.h (hunk 2 above)
Comment (wjakob) [inline include/pybind11/common.h]: Just a ping on this comment.

File: example/example20.cpp (hunk 9 above)
Comment (aldanor) [inline example/example20.cpp]: Yea, good point.

I think in general `py::array` should provide an easier way to get the data pointer. Maybe via array interface and not the buffer interface? (in which case you don't need to request) I could take a look later

File: include/pybind11/common.h (hunk 2 above)
Comment (aldanor) [inline include/pybind11/common.h]: Yep, on my todo list, all comments in this thread will be addressed :)

File: include/pybind11/numpy.h [hunk 10]
```diff
@@ -84,37 +120,33 @@ class array : public buffer {
...
-        if (info.format == "Zd")      fmt = API::NPY_CDOUBLE_;
-        else if (info.format == "Zf") fmt = API::NPY_CFLOAT_;
-
-        PyObject *descr = api.PyArray_DescrFromType_(fmt);
-        if (descr == nullptr)
-            pybind11_fail("NumPy: unsupported buffer format '" + info.format + "'!");
+        auto& api = lookup_api();
+
+        // _dtype_from_pep3118 returns dtypes with padding fields in, however the array
+        // constructor seems to then consume them, so we don't need to strip them ourselves
```
Comment (wjakob) [inline include/pybind11/numpy.h]: I vaguely remember commenting about this but can't find it anymore. Basically the comment above makes me scratch my head: it says "we _don't_ need to strip [.. padding fields ..]" but then the code below calls a strip_padding_fields function. Maybe just a wording issue?

File: include/pybind11/numpy.h [hunk 11]
```diff
@@ -123,6 +155,50 @@ class array : public buffer {
...
+        std::vector<field_descr> field_descriptors;
+
+        auto items = fields.attr("items").cast<object>();
+        for (auto field : items()) {
+            auto spec = object(field, true).cast<tuple>();
+            auto name = spec[0].cast<pybind11::str>();
+            auto format = spec[1].cast<tuple>()[0].cast<object>();
+            auto offset = spec[1].cast<tuple>()[1].cast<int_>();
+            if (!len(name) && (std::string) dtype.attr("kind").cast<pybind11::str>() == "V")
+                    continue;
```
Comment (wjakob) [inline include/pybind11/numpy.h]: wrong indent?

File: example/example20.cpp (hunk 9 above)
Comment (wjakob) [inline example/example20.cpp]: That's a good suggestion -- I would say let's do this via a separate PR though, this one is already getting big enough ;)

**Comment by wjakob** [comment]
`PYBIND11_DESCR` really has to stay `constexpr`. This enables generating function signatures at compile time, which reduces the size of generated binaries considerably.

I am greatly in favor of introducing a py::dtype object that can be used dynamically (in whatever way makes sense -- don't interpret my previous suggestion for an API too strictly).

File: include/pybind11/numpy.h (hunk 10 above)
Comment (aldanor) [inline include/pybind11/numpy.h]: I was wrong, will fix the comment. The reason I thought they get stripped is because of the whole "forcecast is the default" story -- which forced those phantom padding fields to disappear; forgot to update the comment. I've also changed the casting flags to 0 in all tests for this PR.

File: include/pybind11/numpy.h (hunk 11 above)
Comment (aldanor) [inline include/pybind11/numpy.h]: Indeed :)

**Comment by aldanor** [comment]
@wjakob I believe I've addressed all of the comments here (last 6 commits), except moving dtype-related logic inside of a `py::dtype` wrapper -- working on that right now.

I've also recalled I forgot to add descriptor bindings for one important class of types: string fields; it's fairly straightforward actually (e.g. so that `char[7]` and `std::array<char, 7>` map to `S7` in numpy and `7s` in buffer protocol), however I ran into a weird problem with `detail::_`, maybe I don't fully understand how it works :)

``` cpp
template <size_t N> struct npy_format_descriptor<char[N]> {
    static PYBIND11_DESCR name() { return _("S") + _<N>(); } 
    // more stuff
};
```

This compiles/links fine, however if you try to run it with an example, you get

``` sh
Traceback (most recent call last):
  File "../example/example20.py", line 6, in <module>
    from example import (
ImportError: dlopen(../example/example.so, 2): Symbol not found: __ZN8pybind116detail10int_to_strILm0EJLm3EEE6digitsE
  Referenced from: ../example/example.so
  Expected in: flat namespace
 in ../example/example.so
```

So I can't use `_` on template parameters? 🐼

**Comment by wjakob** [comment]
Weird .. sounds like a potential compiler issue. The whole point of `_` is that it can be used with template parameters :)

**Comment by aldanor** [comment]
@wjakob Hmm, a very short example (this is clang on OSX, c++14). Should this work / compile?

``` cpp
#include <pybind11/pybind11.h>

namespace pybind11 {
namespace detail {
template<size_t N> PYBIND11_DESCR foo() { return _("S") + _<N>(); }
}
}

int main() {
    std::cout << pybind11::detail::foo<3>().text() << std::endl;
}
```

``` sh
Undefined symbols for architecture x86_64:
  "pybind11::detail::int_to_str<0ul, 3ul>::digits", referenced from:
      auto pybind11::detail::_<3ul>() in pb11-_-2001d1.o
ld: symbol(s) not found for architecture x86_64
```

(This is built without `-undefined dynamic_lookup` so it actually triggers a compile error and not a runtime one)

Interestingly, it seems to work on gcc 5 on Linux (c++14), just checked.

**Comment by wjakob** [comment]
Fixed (4a87933be9c816b0434f563c263c260a41ed31d0). The issue is that a reference of a constexpr value was taken, which requires it to actually exist at runtime. The commit now switches to pass-by-value at compile time.

**Comment by aldanor** [comment]
Cool, I'll check this tonight when I get back to my mac :+1:

// Wonder why gcc5 was fine with it though?

**Comment by wjakob** [comment]
I think whether this should give a linker error is implementation/optimization-specific behavior that is not covered by the spec.

**Comment by aldanor** [comment]
@wjakob Your `int_to_str` fix seems to have worked, I've implemented format/numpy descriptors for `char[N]` and `std::array<char, N>` + tests (that's not directly recarray-related, but very dtype-related, and is a great addition to our numpy arsenal, idk if it's worth mentioning in the docs or not), works on both clang/osx and gcc5/linux.

I've also moved all numpy API-related stuff out of `py::array` into a separate container for the sake of clarity, it was started to get swamped with it. So now the last cleanup bit is finishing `py::dtype`, hope that won't take long, will keep you posted :)

**Comment by wjakob** [comment]
Ok that sounds good! Please go ahead and create an official PR at the main repo once you have something.

**Comment by aldanor** [comment]
@wjakob This is the final bit of this PR, I've pushed preliminary implementation of `py::dtype` which mostly works (all tests pass, but when the process closes there's a segfault in `handle::dec_ref()`, seems like refcount is off somewhere but I can seem to find our where exactly so far........ -- hopefully I can find it).

While I'm searching for the refcount bug, looking forward to any comments and suggestions re: the API 😃

**Comment by wjakob** [comment]
@aldanor: This looks fantastic. I would suggest one minor reorganization and one addition before wrapping it up:

There is a part of the code which sets up the arguments for the dtype and then calls the dtype constructor.

```
args["names"] = names; args["formats"] = formats; args["offsets"] = offsets;
```

It would make more sense to me if (instead of the static `from_args` function) there was a special dtype _constructor_ which directly takes these lists, i.e.:

```
dtype::dtype(list names, list formats, list offsets, size_t itemsize) 
```

(or something like that, feel free to modify)

The other bit that would be super nice to have is an array constructor which can accept the dynamic dtype:

```
array:array(const dtype &dt, size_t size, const void *ptr);
```

**Comment by aldanor** [comment]
Agreed on having two more constructors, good points! (<s>maybe you can spot the ref count bug too, haha?</s> refcount bug fixed)

Should `from_args` stay too or should it become internal? It accepts a much wider range of things (everything that np.dtype does, and it's a lot if you check the docstring). The only reason it's a static function and not a ctor is that it takes a single `object` which would mix up with object's ctors

**Comment by aldanor** [comment]
@wjakob Okay, done, the refcount bug fixed, plus a few constructors added, see latest commits:
- simplified `buffer_info` constructor for 1-D case
- `dtype` constructor accepting names/formats/offsets/itemsize
- `array` gained a constructor accepting a `dtype`
- all copy/paste eliminated in`array` / `array_t` constructors, added a whole bunch of them to cover most common cases (e.g. if you provide just the shape, strides will be auto-computed)

Basically, it's now possible to do stuff like this:

``` cpp
auto arr1 = py::array_t<int>(10);
auto arr2 = py::array("int32", { 3, 2 });
```

without having to mess with buffer_info in such simple examples.

I could probably fix the existing examples in the docs to use those.

Also, the const qualifier for the data pointer was dropped for array ctors -- I'm not absolutely sure about this yet but it looks a more reasonable default, I think? This will prob need to be revisited as a part of the whole casting rework, but generally we do want to create and export arrays that don't own their data (e.g. wrappers to expose huge vectors in c++), I'd say that'd be one of the most common use cases. It's also possible to add ctors from std::vector, too, but that needs more though re: casting and data ownership (the general numpy convention is that nothing is copied unless you explicitly ask for it, e.g. via `ensurecopy` flag; this would also mean though that such a constructor would be unsafe be default since its data could now reside in some external vector whose lifetime is not tied to the array). Maybe it's best to just have pointers to hint at the fact that it's not 100% safe and you should know what you're doing.
